students/philip_behrend/lesson07/html_render.py

```diff
@@ -0,0 +1,184 @@
+
+"""
+A class-based system for rendering html.
+"""
+
+
+# This is the framework for the base class
+class Element(object):
+
+    tag = 'html'
+    ind = "    "
+
+    def __init__(self, content="", **kwargs):
+        self.content = [content]
+        self.kwargs = kwargs
+
+    def append(self, new_content):
+        self.content.append(new_content)
+
+
+    def render(self, out_file, cur_ind = ""):
+        out_file.write("{}<{} ".format(cur_ind,self.tag))
+        for key, value in self.kwargs.items():
+            out_file.write("{}=\"{}\"".format(key, value))
+        out_file.write(">\n")
+        for item in self.content:
+            try:
+                item.render(out_file,cur_ind+self.ind)
+            except AttributeError:
+                out_file.write(cur_ind+self.ind)
+                out_file.write(item)
+                out_file.write("\n")
+            out_file.write(cur_ind)
+            out_file.write("</{}>\n".format(self.tag))
+
+class Html(Element):
+    tag = 'html'
+
+    def render(self, out_file, cur_ind = ""):
+        out_file.write("<!DOCTYPE html>\n") 
+        out_file.write(self.ind)
+        out_file.write("<{} ".format(self.tag))
+        for key, value in self.kwargs.items():
+            out_file.write("{}=\"{}\"".format(key, value))
+        out_file.write(">\n")
+        for item in self.content:
+            try:
+                item.render(out_file, cur_ind+self.ind)
+            except AttributeError:
+                out_file.write(item)
+                out_file.write("\n")
+        out_file.write(cur_ind)
+        out_file.write("</{}>\n".format(self.tag))
+
+
+class Body(Element):
+    tag = 'body'
+
+class P(Element):
+    tag = 'p'
+
+class Head(Element):
+    tag = 'head'
+
+    def render(self, out_file, cur_ind = ""):
+        out_file.write("<{}>\n".format(self.tag))
+        Meta(charset="UTF-8").render(out_file)
+        for item in self.content:
+            try:
+                item.render(out_file, cur_ind+self.ind)
+            except AttributeError:
+                out_file.write(cur_ind)
+                out_file.write(item)
+        out_file.write("\n</{}>\n".format(self.tag))
+
+
+class OneLineTag(Element):
+    tag = 'title'
+
+    def render(self, out_file, cur_ind = ""):
+        out_file.write("<{}> ".format(self.tag))
+        for item in self.content:
+            try:
+                item.render(out_file)
+            except AttributeError:
+                out_file.write(cur_ind)
+                out_file.write(item)
+        out_file.write(" </{}>\n".format(self.tag))
+
+class Title(OneLineTag):
+    tag = 'title'
+
+class SelfClosingTag(Element):
+    tag = 'hr'
+
+    def __init__(self, content=None, **kwargs):
+        self.content = content
+        if self.content is not None:
+            raise TypeError
+        self.kwargs = kwargs
+
+
+    def render(self,out_file,cur_ind=""):
+        out_file.write("<{} ".format(self.tag))
+        if self.kwargs != {}:
```

Comment:
no need for this check, just loop through the dict and if dict is empty then it won't do anything

Reply:
@natasha-aleksandrova 
Hi Natasha, the render function varies for each tag (for example, the HTML render has a DOCTYPE at the top while Head has a charset argument). How can these nuances be captured without fully rewriting each render function?

Reply:
@pbehrend great question and this is what makes inheritance tricky and great at the same time 🙂 

for `HTML` class in the instructions:
> Update the Html element class to render the “<!DOCTYPE html>” tag at the head of the page, before the html element.
You can do this by subclassing Element, overriding render(), but then calling the Element render from the new render.

which means your code could look something like this:
```
def render(self, file_out, cur_ind=""):
        file_out.write(cur_ind + "<!DOCTYPE html>\n")
        super().render(file_out, cur_ind=cur_ind)
```


`charset` is an attribute and should just be handled as part of the `Element` attribute handling.